### ISSUE TITLE
Cannot locate test source when test is inherited from a class in another project

### ISSUE THREAD

**Issue by aceoft** [xunit/xunit#3023]
When inheriting a test from a class in another project (even in the same solution), the source code location for that test is not available to the Visual Studio Test Explorer. I have tested with both abstract and non-abstract base classes.

I am led to believe this is an issue with xunit and not the VS Test Explorer because when using MSTest, the source link is available as expected.

I have searched the issue archive to make sure this isn't a duplicate. I found one issue that looked similar (https://github.com/xunit/xunit/issues/1401) but it was specifically to do with generic classes.

I have reproduced the issue in a sample repo here: https://github.com/aceoft/XUnitInheritedTests

### MSTest
![image](https://github.com/user-attachments/assets/6c79bb65-25b1-4b1c-abe0-458b3eef6ef1)
_The source (`AbstractTests.cs line 7`) is located._

### xUnit V2
![image](https://github.com/user-attachments/assets/3618dcf1-3887-421e-b148-17ff737f1c5e)
_The source is not available._

### xUnit V3
![image](https://github.com/user-attachments/assets/0fff3ead-453b-4745-b958-bd2cb902fe06)
_The source is not available._

Both V2 and V3 do successfully find inherited test sources as long as the tests are in the same project.

![image](https://github.com/user-attachments/assets/f374b1b0-aa02-4c0e-8fc5-487e96e216dd)

Aside from working within the same solution, it would be extremely helpful if this would work as well when the abstract tests are used via a NuGet package (reused across many projects in our organization). I expect that is a new set of issues to work through, but I thought I would mention it.

Thank you for your time and consideration!

**Comment by bradwilson** [comment]
Thanks. We rely on [`DiaSession` from VSTest](https://github.com/microsoft/vstest/blob/main/src/Microsoft.TestPlatform.ObjectModel/Navigation/DiaSession.cs) to do all that mapping for us, so I'll see if there is a way that we can make this work.

(As an aside, it would be interesting to see if this works with the new [Microsoft Testing Platform support](https://xunit.net/docs/getting-started/v3/microsoft-testing-platform) that we're about to ship, because in that scenario we don't do any source mapping at all. I'll make sure to test that and see what happens there.)

As for differences between us and MSTest... I'm not sure whether MSTest also relies on `DiaSession` or not; they may have access to some private APIs that we do not, or they may do the source mapping themselves rather than relying on VSTest. A better test would be to see whether NUnit exhibits the same behavior, as I'm assuming they too are relying on `DiaSession` for the mapping behavior.

I just want to set the expectation that it's possible that we won't be able to fix this (or, if it works w/ Microsoft Testing Platform, it'll only work with v3 + MTP).

**Comment by bradwilson** [comment]
I can verify that using Microsoft Testing Platform mode (rather than VSTest mode) does indeed correctly map the source. Screenshot using VS 2022 17.12 Preview 2:

![image](https://github.com/user-attachments/assets/8d81353c-7a49-4e66-bbce-e2173bcaf3d3)

**Comment by aceoft** [comment]
Thank you for looking into this!

**Comment by bjornhellander** [comment]
I was just going to report the same thing :-) Would be awesome if this could be fixed, since has been bugging me for a coupe of years now.

**Comment by bradwilson** [comment]
Unfortunately I haven't gotten any response yet from VSTest. Maybe add an upvote or some comments on the issue linked above?

**Comment by bjornhellander** [comment]
I posted a hypothesis in the testfx issue after having a look at the MSTest code. Don't know what would be needed here, if it's correct.

**Comment by nohwnd** [comment]
Looking at your repro, in neither case the source information are picked up by DiaSession from PDB. In VS for xunit, nunit and mstest the collection of source information from PDBs is disabled because a more performant source based method location lookup is used. (CollectSourceInformation runsetting is set to false, bridge does it here: https://github.com/microsoft/testfx/blob/main/src/Platform/Microsoft.Testing.Extensions.VSTestBridge/ObjectModel/RunSettingsPatcher.cs#L69).

When I try this and enable Test explorer verbose logs I see this: 

[23/09/2024 12:47:55.317 pm]  Source location lookup for Source: S:\c\XUnitInheritedTests\ProjectB.V3\bin\Debug\net8.0\ProjectB.V3.dll by FullyQualifiedName: ProjectB.V3.DerivedTestsInDifferentProject.InheritedTest
[23/09/2024 12:48:00.137 pm]  Source location lookup for Source: S:\c\XUnitInheritedTests\ProjectB.V2\bin\Debug\net8.0\ProjectB.V2.dll by FullyQualifiedName: ProjectB.V2.DerivedTestsInDifferentProject.InheritedTest
[23/09/2024 12:48:02.014 pm]  Source location lookup for Source: S:\c\XUnitInheritedTests\ProjectB.V2\bin\Debug\net8.0\ProjectB.V2.dll by FullyQualifiedName: ProjectB.V2.DerivedTestsInDifferentProject.DirectTest
[23/09/2024 12:48:02.741 pm]  Source location lookup for Source: 
S:\c\XUnitInheritedTests\ProjectB.V2\bin\Debug\net8.0\ProjectB.V2.dll by FullyQualifiedName: ProjectB.V2.DerivedTestsInDifferentProject.InheritedTest
[23/09/2024 12:48:06.765 pm]  Source location lookup for Source: S:\c\XUnitInheritedTests\ProjectB.MSTest\bin\Debug\net8.0\ProjectB.MSTest.dll by ManagedType: ProjectB.MSTest.DerivedTestsInDifferentProject and ManagedMethod: InheritedTest


Where mstest tests are discovered using managed name and managed method, while xunit tests are looked up by fully qualified name, which points to the class that inherited the method rather than the "real" placement of the method. 

I don't think this is a new bug, but rather a limitation of the fullyqualified name. But please correct me if  I am wrong. 

I cannot duplicate your run via the new testing platform, but I assume we are setting the managed method property there, and that is why the correct method is picked up from the VS metadata it collected. 

@drognanar to correct me if I am wrong about how TE uses the FQN and the new managed method and managed type properties on tests.

**Comment by nohwnd** [comment]
> I cannot duplicate your run via the new testing platform, but I assume we are setting the managed method property there, and that is why the correct method is picked up from the VS metadata it collected.

I can duplicate it, I don't know why I thought you did not release it publicly yet. ManagedType and method name is used for the lookup and in that case TE works correctly. Even though I see that the type points to ProjectB dll not ProjectA dll so some more lookups are probably done on TE side.

**Comment by bradwilson** [comment]
> CollectSourceInformation runsetting is set to false, bridge does it here

We don't use the bridge. Is this a value we should be setting?

> Source location lookup for Source: S:\c\XUnitInheritedTests\ProjectB.MSTest\bin\Debug\net8.0\ProjectB.MSTest.dll by ManagedType: ProjectB.MSTest.DerivedTestsInDifferentProject and ManagedMethod: InheritedTest

Is this calling a different DiaSession API?

**Comment by nohwnd** [comment]
In that case you are probably not collecting source information at all, which is imho fine, since visual studio client disables it by default. And commandline scenarios don’t use it. At least in the diag logs I looked at there is no source information logged. 

If you are not using bridge then you are setting the TestNode values yourself, and you are setting them correctly to make TE work correctly. 



The source location lookup entry is written by TE, and refers to looking up source based discovery data based on the type. It is not related to DiaSession or any other PDB based source data lookup. Sorry for being confusing.

**Comment by bradwilson** [comment]
My problem isn't with MTP (which calls DiaSession--or something like it--on my behalf), my problem is with VSTest (where I have to call DiaSession myself) and understanding why one works and the other doesn't.

**Comment by nohwnd** [comment]
@bradwilson I am trying to explain that you are NOT calling DiaSession in either case (as far as I can see). 🙂

The location information is coming from VS scanning the source code via Roslyn and picking up location of `[Fact]` and `[Theory]` attributes, and linking them to methods and types.

The difference is that with xunit2 + VSTest the test has FullyQualifiedName, for which the TE does not have location: 

(this is reported by TestExplorer)

```
[24/09/2024 5:51:37.455 pm]  Source location lookup for Source: S:\c\XUnitInheritedTests\ProjectB.V2\bin\Debug\net8.0\ProjectB.V2.dll 
by FullyQualifiedName: <---
ProjectB.V2.DerivedTestsInDifferentProject.InheritedTest
```

![image](https://github.com/user-attachments/assets/475d3287-dc1c-472e-a304-bc5823ed9fad)

---

For with xunit3 + MTP it has additional ManagedMethod and ManagedType properties (that are passed on TestNode), that are a newer way of telling TE where the type is, and for that TE is able to find the location:

```
[24/09/2024 5:52:34.626 pm]  Source location lookup for Source:
S:\c\XUnitInheritedTests\ProjectB.V3\bin\Debug\net8.0\ProjectB.V3.dll
by ManagedType: ProjectB.V3.DerivedTestsInDifferentProject  <---
and ManagedMethod: InheritedTest <---
```

![image](https://github.com/user-attachments/assets/8ca67ea3-f427-4b80-a13a-f8b02857c0d7)

--- 

MSTest is using managed type, managed method properties since 2.x.x and so it also works there, even if you use it with vstest: 

```
[24/09/2024 5:55:19.235 pm]  Source location lookup for Source: S:\c\XUnitInheritedTests\ProjectB.MSTest\bin\Debug\net8.0\ProjectB.MSTest.dll 
by ManagedType: ProjectB.MSTest.DerivedTestsInDifferentProject  <---
and ManagedMethod: InheritedTest <---
```

 
![image](https://github.com/user-attachments/assets/b9d26bfb-83ba-49c9-834a-a0179b797425)

---

So if this worked and does not work anymore this is a bug in TestExplorer (please report it via VS feedback). If it did not work before, it is a limitation of the FQN approach, and TE will tell you to send the new ManagedMethod, ManagedType and Hierarchy properties. 

https://github.com/microsoft/testfx/blob/main/src/Adapter/MSTest.TestAdapter/Extensions/TestCaseExtensions.cs#L16-L36

@drognanar do we have those documented with examples somewhere?

**Comment by bradwilson** [comment]
> @bradwilson I am trying to explain that you are NOT calling DiaSession in either case (as far as I can see). 🙂

We most certainly are, via `VisualStudioSourceInformationProvider`. Whether we need to or not any more, is a separate question. It sounds like Test Explorer was updated at some point to call DiaSession (or its equivalent) if we don't pass source information, but that was not always the case (or we would've never written any of this DiaSession code).

> @drognanar do we have those documented with examples somewhere?

Yes, please, point to the documentation. Thanks!

**Comment by bradwilson** [comment]
To test this, I've temporarily created an inherited test situation, with one local method and three inherited methods.

Test 1: I remove my calls to DiaSession, nothing else. Result: only get source for the local method.

https://github.com/user-attachments/assets/31dc41de-691c-40b1-b8e2-a25d30db3142

Test 2: I add in ManagedType and ManagedMethod, I get source for [Fact]s but not for [Theory]s with pre-enumerated data. Note: I have verified that both test cases for `ConditionalSkip` are correctly setting the type name and method name.

https://github.com/user-attachments/assets/cd93f2c9-b6d3-4eac-a216-bf52141a2e19

So to me, this looks better, but not 100% fixed.

**Comment by bradwilson** [comment]
Note that it does perform 100% correctly in Microsoft Testing Platform mode (this is VS2022 17.12 Preview 2). Obviously this bypasses all VSTest code, so `xunit.runner.visualstudio` isn't involved at all here (and you can tell it's MTP mode because the tree structure is different).

https://github.com/user-attachments/assets/f96bab71-5172-4bd5-b0f7-b4d0a0387097

**Comment by bradwilson** [comment]
To test the parameter type postulation, I turned off theory pre-enumeration back in VS2022 17.11 (so, back in VSTest mode) and still no source information there, either:

![image](https://github.com/user-attachments/assets/b5ca136b-6aef-43ad-ba7b-65c7979eb9ea)

That strongly suggests to me that the parameter type information is important to this process.

**Comment by bradwilson** [comment]
Yep, parameter types are required. I added a second parameter to make sure a simple comma-joined list was what it was after:

![image](https://github.com/user-attachments/assets/0d004aa1-694f-43a3-b218-95e3a45f1054)

I'm not sure whether removing my `DiaSession` code is a good idea because of v1/v2 tests not reporting their parameter types (this was newly added to v3 to support MTP), so I'm still poking around to see.

**Comment by bradwilson** [comment]
Still not working 100%, and it's broken with generic types.

I have this as a signature:

`public void AppDomain(string value, AppDomainSupport? expected)`

And this is what I report as the managed method name:

`AppDomain(System.String,System.Nullable`1[[Xunit.Runner.Common.AppDomainSupport, xunit.v3.runner.common, Version=0.4.0.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]])`

That shows no source information.

**Comment by bradwilson** [comment]
If I take away the parameter information and just report the method name as `AppDomain`, then it works:

![image](https://github.com/user-attachments/assets/3f0053dd-0236-4a03-add6-55ba4e9e19f5)

I'm very confused by the logic that's being used here, because of the lack of consistency.

**Comment by bradwilson** [comment]
I guess, based on what I've learned, I should put parameter types on the method name _**unless**_ the parameter types include generics. That's probably the way I get the best coverage.

Hoping documentation clears up exactly what's going on here and why there's this inconsistency between local vs. cross-project lookups. @drognanar

**Comment by bradwilson** [comment]
I'm really extra confused now.

For this:

`public void AppDomain(string value, AppDomainSupport? expected)`

These work (including the fact that one of them is just...wrong?):

```AppDomain(System.String,System.Nullable`1[[Xunit.Runner.Common.AppDomainSupport]])```
```AppDomain(System.String,System.Nullable`1)```
```AppDomain(System.String)```

The only thing that seems to break it is the `[[...]]` including commas.

I can't come up with an explanation for why it's broken when it shouldn't be, and working when it shouldn't be.

**Comment by bradwilson** [comment]
@aceoft You can try `3.0.0-pre.37` (https://xunit.net/docs/using-ci-builds) and see if that doesn't solve the problem for you (mostly). The place where it's still known to be broken would be any test method, cross-project, which has a parameter type that is generic (which includes nullable value types, like `int?`, since they surface as `Nullable<int>`).

**Comment by bradwilson** [comment]
I'm opening a new issue about generics not working, because:

(a) if that gets fixed, I can always report the managed method with its parameter types
(b) there's no way for me to "fix" this in Microsoft Testing Platform mode, as I am not responsible for constructing those managed method properties (MTP is)

https://github.com/microsoft/testfx/issues/3863

**Comment by bradwilson** [comment]
Final fix (including generics) is available in v3 `0.5.0-pre.12` and xunit.runner.visualstudio `3.0.0-pre.38`.
